### DISCUSSION TITLE
Sobre uso de Multiplexer

### DISCUSSION THREAD

**Discussion by alecmg** [MobiFlight/MobiFlight-Connector#441] [General]
Good Morning. Sorry I'm new here, I would like to know if you intend to add the use of multiplexers of the type Cd74hc4067 with the arduino for expansion of digital and analog ports. Thanks and congratulations for the work

**Comment by ismailsurek** [comment]
multiplexers of the type Cd74hc4067  is possible to use it with LUA script PMDG using 
for arduino 
#include "Type4067Mux.h"
Type4067Mux mux1(8, INPUT, DIGITAL, 12, 11, 10, 9);  ----- ( Sig= 8 , S0 = 12,  S1=11,  S2=10 ,  S3=9)
Type4067Mux mux2(3, INPUT, DIGITAL, 12, 11, 10, 9);   ----  ( Sig=3 ,   S0 =12,  S1=11,  S2=10 ,  S3=9)
// all s0 s1 s2 s3 bus common connect
// you can extend this line to how many do you want as arduino suppy .. Only change SIG pin of arduino 

byte data;
byte i;

int M1[16] = {false,false,false,false,false,false,false,false,false,false,false,false,false,false,false};
int M1_old[16] = {false,false,false,false,false,false,false,false,false,false,false,false,false,false,false};
int M1_list[16] = {false,false,false,false,false,false,false,false,false,false,false,false,false,false,false};

int M2[16] = {false,false,false,false,false,false,false,false,false,false,false,false,false,false,false};
int M2_old[16] = {false,false,false,false,false,false,false,false,false,false,false,false,false,false,false};
int M2_list[16] = {false,false,false,false,false,false,false,false,false,false,false,false,false,false,false};


void setup() {
  Serial.begin(115200);
  while (!Serial);
}

void check_status(){
  for (i = 0; i < 16; i++) {
    data = mux1.read(i);
    (data == HIGH) ? (M1[i] = true) : (M1[i] = false);
  }
  for (i = 0; i < 16; i++) {
    data = mux2.read(i);
    (data == HIGH) ? (M2[i] = true) : (M2[i] = false);
  }  

void monitor_status(){
  Serial.println("-------------------");
  for (i = 0; i < 16; i++) {
    if (M1[i] == true) {
      Serial.println("TRUE");
    } 
    else if (M1[i] == false) {
      Serial.println("FALSE");
    }
  }
  for (i = 0; i < 16; i++) {
    if (M2[i] == true) {
      Serial.println("TRUE");
    } 
    else if (M2[i] == false) {
      Serial.println("FALSE");
    }
  }
  Serial.println("-------------------");
}

void diff_detector(){
  for (i = 0; i < 16; i++) {
    (M1[i] != M1_old[i]) ? (M1_list[i] = true) : (M1_list[i] = false);
    (M2[i] != M2_old[i]) ? (M2_list[i] = true) : (M2_list[i] = false);
}

void monitor_diff(){
  for (i = 0; i < 16; i++) {
    if(M1_list[i] == 1){
      Serial.println("DEGISTI");
    }
  }
  for (i = 0; i < 16; i++) {
    M1_list[i] = 0;
  }
  for (i = 0; i < 16; i++) {
    if(M2_list[i] == 1){
      Serial.println("DEGISTI");
    }
  }
  for (i = 0; i < 16; i++) {
    M2_list[i] = 0;
  }
  //Serial.println("monitor_diff_done");
}

void send_diff(){
  for (i = 0; i < 16; i++) {
    if(M1_list[i] == true){
      (M1[i] == true) ? (Serial.println("P" + String(i))) : (Serial.println("R" + String(i)));
    }
    if(M2_list[i] == true){
      (M2[i] == true) ? (Serial.println("P" + String(i+16))) : (Serial.println("R" + String(i+16)));
    }

void monitor_diff(){
  for (i = 0; i < 16; i++) {
    if(M1_list[i] == 1){
      Serial.println("DEGISTI");
    }
  }
  for (i = 0; i < 16; i++) {
    M1_list[i] = 0;
  }
  for (i = 0; i < 16; i++) {
    if(M2_list[i] == 1){
      Serial.println("DEGISTI");
    }
  }
  for (i = 0; i < 16; i++) {
    M2_list[i] = 0;
  }
  //Serial.println("monitor_diff_done");
}

void send_diff(){
  for (i = 0; i < 16; i++) {
    if(M1_list[i] == true){
      (M1[i] == true) ? (Serial.println("P" + String(i))) : (Serial.println("R" + String(i)));
    }
    if(M2_list[i] == true){
      (M2[i] == true) ? (Serial.println("P" + String(i+16))) : (Serial.println("R" + String(i+16)));
}  
void shifter(){
  for (i = 0; i < 16; i++) {
    M1_old[i] = M1[i];
    M2_old[i] = M2[i];
void loop() {
} 

For Lua script ;
example code Function ;
dev = com.open("COM5", 9600, 0) --

if dev == 0 then -- 
    ipc.display("Could not open device port")
    ipc.exit()
end
function pkgbrake(offset, value)
    if value == 1 then
        com.write(dev, "T000\n")
    else
        com.write(dev, "S000\n")
    end
end

event.com(Serial_Com_Port, 50,1, "Serial_Data")


--[[  FMC icin
	if (string.find(datastring, "Pxxx")) then 
		ipc.control(xxxxx,1)  
		print(datastring)
  end  
]]

**Comment by alecmg** [comment]
cool thank you very much!

**Comment by MobiFlight-Admin** [comment]
Hey everyone, I had a story for this on the backlog for a longer while. With the addition of Shift Output and Input Registers we have enhanced the UI already. I would also like to add this particular chip to the list of supported devices. So please keep an eye on #276 for progress.
